packages/nimble-components/src/rich-text/editor/index.ts

```diff
@@ -262,8 +273,10 @@ export class RichTextEditor extends RichText implements ErrorPattern {
             this.richTextMarkdownSerializer = new RichTextMarkdownSerializer(
                 mentionExtensionConfig.map(config => config.name)
             );
+            const formatButtonsState = this.getButtonsState(this.tiptapEditor);
             this.initializeEditor();
             this.setMarkdown(currentStateMarkdown);
+            this.resetEditorButtonsState(formatButtonsState);
         }
         this.setActiveMappingConfigs();
```

Comment:
It looks like every code path through configurationChanged ends up saving and restoring the button state.

Instead of duplicating the logic in two places can we write it that way? Also I don't think having the separate `refreshMarkdownContent` function is making this code clearer. They duplicate so much functionality it does not seem helpful.

```ts
configurationChanged () {
    const formatButtonsState = this.getButtonsState(this.tiptapEditor);
    if (this.isMentionExtensionConfigUnchanged(prev, next)) {
          const { from, to } = this.tiptapEditor.view.state.selection;
          this.setMarkdown(this.getMarkdown());
          this.tiptapEditor.commands.setTextSelection({ from, to });
    } else {
            const mentionExtensionConfig = this.getMentionExtensionConfig();
            const currentStateMarkdown = this.getMarkdown();
            this.richTextMarkdownSerializer = new RichTextMarkdownSerializer(
                mentionExtensionConfig.map(config => config.name)
            );
            this.initializeEditor();
            this.setMarkdown(currentStateMarkdown);
    }
    this.resetEditorButtonsState(formatButtonsState);
    this.setActiveMappingConfigs();
}

```

Also the above makes me wonder, how come in the else case we are not saving the selection state? Should we do it in both cases like the following?

```ts
configurationChanged () {
    const formatButtonsState = this.getButtonsState(this.tiptapEditor);
    const { from, to } = this.tiptapEditor.view.state.selection;
    if (this.isMentionExtensionConfigUnchanged(prev, next)) {
          this.setMarkdown(this.getMarkdown());
    } else {
            const mentionExtensionConfig = this.getMentionExtensionConfig();
            const currentStateMarkdown = this.getMarkdown();
            this.richTextMarkdownSerializer = new RichTextMarkdownSerializer(
                mentionExtensionConfig.map(config => config.name)
            );
            this.initializeEditor();
            this.setMarkdown(currentStateMarkdown);
    }
    this.tiptapEditor.commands.setTextSelection({ from, to });
    this.resetEditorButtonsState(formatButtonsState);
    this.setActiveMappingConfigs();
}

```

If not can we add a comment about why selection state needs to be saved in one and not the other?

Reply:
Added `resetEditorButtonsState()` in both the states.